### ISSUE TITLE
Some decision tree branching may be unnecessary

### ISSUE THREAD

**Issue by kevinrobinson** [IBM/taxinomitis#226]
@dalelane https://github.com/IBM/taxinomitis/pull/221 is amazing, awesome work! 👍  Thanks as always for sharing such great work in the open! ❤️ 

I made some limited test data just to check this out, and noticed that the explanation had some parts of the tree that didn't seem necessary since they all result in the same classification.  Here's an example, with the parts highlighted that seem unnecessary:

![image](https://user-images.githubusercontent.com/1056957/63272294-ce4cb900-c269-11e9-8ea2-3d684c623132.png)

Maybe the way the tree is constructed can lead to this?  I'm not sure.  I tried to see where the tree is constructed, guessing that this was done by a service but only got as far as https://github.com/IBM/taxinomitis/blob/db15bf81260149efeb28a29d2e0900580bfb6406/src/lib/training/numbers.ts#L328 and wasn't quite sure from there.

I'm not sure if this is an actual issue, as in something isn't working as it should, or just something that might be worth adding to your great explanation up top.

Also, http://www.r2d3.us/visual-intro-to-machine-learning-part-1/ has some awesome visuals of decision trees that might be a good fit as a "learn more" link in the explanation up to as well.

**Comment by dalelane** [comment]
I'll have a look, thanks!

In the meantime, the code where this is done lives here:
https://github.com/IBM/taxinomitis/blob/db15bf81260149efeb28a29d2e0900580bfb6406/serverless-functions/mltraining-numbers/numbers.py#L43-L56

**Comment by kevinrobinson** [comment]
@dalelane awesome, thanks!

Yeah, it seems like the scikit learn algorithm is an approximation since for larger real-world data sets it can be intractable to compute the optimal representation [link](https://scikit-learn.org/stable/modules/tree.html):

> The problem of learning an optimal decision tree is known to be NP-complete under several aspects of optimality and even for simple concepts. Consequently, practical decision-tree learning algorithms are based on heuristic algorithms such as the greedy algorithm where locally optimal decisions are made at each node. Such algorithms cannot guarantee to return the globally optimal decision tree. This can be mitigated by training multiple trees in an ensemble learner, where the features and samples are randomly sampled with replacement.

I suppose to simplify things for folks it'd be possible to write a function that does that simplification on the tree that's returned, since for ML for kids models the trees wouldn't be expensive to prune after the fact.  sci-kit learn exposes the internals of the decision tree structure ([link](https://scikit-learn.org/stable/auto_examples/tree/plot_unveil_tree_structure.html#sphx-glr-auto-examples-tree-plot-unveil-tree-structure-py)) but the graphviz function takes a `dt` so that might involve hacking rather than just walking and modifying the tree.  Alternately it might be worth just adding a line of explanation alerting learners and teachers to this, and calling it out as an example of the kind of approximation that is common in ML.

Relatedly, it looks like subsequent runs are non-deterministic [link](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html):

> The features are always randomly permuted at each split. Therefore, the best found split may vary, even with the same training data and max_features=n_features, if the improvement of the criterion is identical for several splits enumerated during the search of the best split. To obtain a deterministic behaviour during fitting, random_state has to be fixed.

This might be good to either add a line about in the explanation, or to avoid this to simplify the experience and fix a random seed so that folks aren't confused if they get different trees from different training runs on the same data set (eg, within a class).

**Comment by dalelane** [comment]
Related issues:
https://github.com/scikit-learn/scikit-learn/issues/10810
https://github.com/scikit-learn/scikit-learn/issues/6557
https://github.com/scikit-learn/scikit-learn/issues/4630
